File: contest/joined/TOPC-2023-PRE/H.cpp
// Author : ysh
// 2023/09/16 Sat 10:52:41
#include<bits/stdc++.h>
using namespace std;
#ifdef LOCAL
#include<debug.h>
#else
#define debug(...) '*'
#define printf(...) '*'
#endif

#define sig(i) ((1 << i) - 1)
int siginv(int x) {
    int now = 1;
    int ans = 0;
    while(x > 0) {
        x = x - now;
        now = now << 1;
        ans++;
    }
    return ans;
}
#define int unsigned long long
signed main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int a,b;cin>>a>>b;

    int height = siginv(a);
    #define h height
    int left = sig(h) - a;
    
    int ans = 0;
    debug(height,left);
    function<void(int)> check = [&] (int x) {
        int larger = sig(h - (x - 1)) - min(left,(unsigned long long) (1 << (h - x))) - 1;
        int smaller = x - 1;
        if(smaller >= b || x > h) return;
        if(b >= smaller + 1 and b <= a - larger) {
            ans = ans + (1 << (x - 1));
        }
        debug(x,smaller + 1,a - larger);
        check(x + 1);
        return;
    };

    check(1);

    cout<<ans;
    return 0;
}

File: contest/joined/TOPC-2023-PRE/H.py
# Author : ysh
# 2023/09/16 Sat 11:11:51
def sig(i):
    return (2 ** i) - 1

def siginv(x):
    now = 1
    ans = 0
    while x > 0:
        x = x - now
        now = now <<1
        ans = ans + 1
    return ans

a,b = [int(i) for i in input().split()]
h = height = siginv(a)
left = sig(h) - a

ans = 0

for x in range(1,h + 1):
    larger = sig(h - (x - 1)) - min(left,(2 ** (h - x))) - 1
    smaller = x - 1
    if smaller >= b or x > h:
        continue
    if b >= smaller + 1 and b <= a - larger:
        ans = ans + (2 ** (x - 1))
    # print(smaller + 1,a - larger)

print(ans)
quit()